src/saveload/saveload.cpp

```diff
@@ -2301,105 +2305,266 @@ struct LZMASaveFilter : SaveFilter {
+		}
+	}
+
+	/** Clean up what we allocated. */
+	~ZSTDSaveFilter()
+	{
+		ZSTD_freeCCtx(this->zstd);
+	}
+
+	/**
+	 * Helper loop for writing the data.
+	 * @param p             The bytes to write.
+	 * @param len           Amount of bytes to write.
+	 * @param end_directive End directive for ZSTD_compressStream2.
+	 */
+	void WriteLoop(byte *p, size_t len, ZSTD_EndDirective end_directive)
+	{
+		byte buf[MEMORY_CHUNK_SIZE]; // output buffer
+		ZSTD_inBuffer input { p, len, 0 };
+
+		size_t ret;
+		do {
+			ZSTD_outBuffer output { buf, sizeof(buf), 0 };
+
+			ret = ZSTD_compressStream2(this->zstd, &output, &input, end_directive);
+			if (ZSTD_isError(ret)) SlError(STR_GAME_SAVELOAD_ERROR_BROKEN_INTERNAL_ERROR, "zstd returned error code");
+
+			if (output.pos != 0) this->chain->Write(buf, output.pos);
+		} while (end_directive == ZSTD_e_end ? (ret != 0) : (input.pos != input.size));
+	}
+
+	void Write(byte *buf, size_t size) override
+	{
+		this->WriteLoop(buf, size, ZSTD_e_continue);
+	}
+
+	void Finish() override
+	{
+		this->WriteLoop(nullptr, 0, ZSTD_e_end);
+		this->chain->Finish();
+	}
+};
+
+#endif /* WITH_LIBZSTD */
+
 /*******************************************
  ************* END OF CODE *****************
  *******************************************/
 
 /** The format for a reader/writer type of a savegame */
 struct SaveLoadFormat {
+	uint8 index;                          ///< Index of this entry.
+	uint8 priority_persistent;            ///< The best (in balance of speed vs size) to use for persistent savegames, those that should survive for a long time and work on other devices. (higher = better)
+	uint8 priority_temporary;             ///< The best (in balance of speed vs size) to use for temporary savegames, like network games. (higher = better)
+
 	const char *name;                     ///< name of the compressor/decompressor (debug-only)
 	uint32 tag;                           ///< the 4-letter tag by which it is identified in the savegame
 
 	LoadFilter *(*init_load)(LoadFilter *chain);                    ///< Constructor for the load filter.
 	SaveFilter *(*init_write)(SaveFilter *chain, byte compression); ///< Constructor for the save filter.
 
-	byte min_compression;                 ///< the minimum compression level of this format
-	byte default_compression;             ///< the default compression level of this format
-	byte max_compression;                 ///< the maximum compression level of this format
+	int8 min_compression;                 ///< the minimum compression level of this format
+	int8 default_compression;             ///< the default compression level of this format
+	int8 max_compression;                 ///< the maximum compression level of this format
 };
 
 /** The different saveload formats known/understood by OpenTTD. */
-static const SaveLoadFormat _saveload_formats[] = {
+static const std::vector<SaveLoadFormat> _saveload_formats = {
```

Comment:
why change it to a vector? at least it could be a std::array

Reply:
I wanted to make this part of the code more C++-like, as the way `_saveload_formats` was used, was annoying to use (iterate over etc).

I changed it in a `std::array` now; not sure what I have to think about it myself, but I would love suggestions to make it better / do it differently :)